### PR TITLE
Fail HTLC backwards before upstream claims on-chain

### PR DESCRIPTION
**Pull request by alecchendev**
Fail inbound HTLCs if they expire within a certain number of blocks from the current height. If we haven't seen the preimage for an HTLC by the time the previous hop's timeout expires, we've lost that inbound HTLC, so we might as well fail it back instead of having our counterparty force-close the channel.

In the occasion that we both 1) see our downstream timeout HTLC confirm and 2) we pass the height where we want to fail back before the upstream expires, we add a check to avoid duplicate events that would lead to failing back the same HTLC twice.

Addresses #2275.

### CODE REVIEW

File: lightning/src/chain/channelmonitor.rs [hunk 1]
```diff
@@ -3353,20 +3416,24 @@ impl<Signer: WriteableEcdsaChannelSigner> ChannelMonitorImpl<Signer> {
...
-						htlc_value_satoshis,
-					}));
-					self.htlcs_resolved_on_chain.push(IrrevocablyResolvedHTLC {
-						commitment_tx_output_idx,
-						resolving_txid: Some(entry.txid),
-						resolving_tx: entry.transaction,
-						payment_preimage: None,
-					});
+					let duplicate_event = self.pending_monitor_events.iter().any(
+						|update| if let &MonitorEvent::HTLCEvent(ref upd) = update { upd.source == *source } else { false });
```
Comment (dunxen) [inline lightning/src/chain/channelmonitor.rs:3420]: Does this check for duplicate events belong in this commit?

File: lightning/src/ln/functional_tests.rs [hunk 2]
```diff
@@ -2211,6 +2211,58 @@ fn channel_reserve_in_flight_removes() {
 	claim_payment(&nodes[0], &[&nodes[1]], payment_preimage_3);
 }
 
+#[test]
+fn test_fail_back_before_backwards_timeout() {
+	// Test that we fail an HTLC upstream if we are still waiting for confirmation downstream
+	// by the time the upstream timeout expires
```
Comment (dunxen) [inline lightning/src/ln/functional_tests.rs:2217]: "just before" it expires?

File: lightning/src/chain/channelmonitor.rs (hunk 1 above)
Comment (alecchendev) [reply lightning/src/chain/channelmonitor.rs:3420]: Before this commit we never had to check for duplicates since this was the only place we pushed the event and we wouldn't repeat it AFAICT, but this commit introduced a new place where we're pushing this event so that's why I included it in the same one. I could separate it out into it's own commit? I don't feel strongly either way so let me know

Comment (dunxen) [reply lightning/src/chain/channelmonitor.rs:3420]: Oh sorry, missed that. Okay definitely belongs here.

File: lightning/src/ln/functional_tests.rs [hunk 3]
```diff
@@ -2211,6 +2211,58 @@ fn channel_reserve_in_flight_removes() {
...
+			assert!(update_fee.is_none());
+			(update_fail_htlcs[0].clone(), commitment_signed.clone())
+		},
+		_ => panic!("Unexpected event"),
+	};
+
+	nodes[0].node.handle_update_fail_htlc(&nodes[1].node.get_our_node_id(), &update_fail);
+	commitment_signed_dance!(nodes[0], nodes[1], commitment_signed, false);
+	expect_payment_failed_conditions(&nodes[0], payment_hash_1, false, PaymentFailedConditions::new().blamed_chan_closed(true));
+}
```
[the remark this answers is not in the document]
Comment (alecchendev) [reply lightning/src/ln/functional_tests.rs:2292]: Ah! Extending this caught a bug - I realized I had covered the case of a duplicate fail within the same block scan, but not across multiple. Added a new commit to prevent this.

**Comment by alecchendev** [comment]
Added one more to do in the PR description, will make a comment when I complete it

**Comment by alecchendev** [comment]
Added last test cases to handle getting update_fail/fulfill_htlc after failing back. Once we get to a point where we fail back the backwards channel, we've already force closed our forwards channel with our timeout, so we just handle these messages as if the channel doesn't exist, which makes sense to me.

**Comment by alecchendev** [comment]
Squashed + rebased

File: lightning/src/chain/channelmonitor.rs [hunk 4]
```diff
@@ -3303,6 +3312,60 @@ impl<Signer: WriteableEcdsaChannelSigner> ChannelMonitorImpl<Signer> {
...
+					let cltv_expiry = match source {
+						HTLCSource::PreviousHopData(HTLCPreviousHopData { cltv_expiry: Some(cltv_expiry), .. }) => *cltv_expiry,
+						_ => continue,
+					};
+					if cltv_expiry <= height + TIMEOUT_FAIL_BACK_BUFFER {
+						let duplicate_event = self.pending_monitor_events.iter().any(
+							|update| if let &MonitorEvent::HTLCEvent(ref upd) = update {
+								upd.source == *source
+							} else { false });
+						if !duplicate_event {
```
[the remark this answers is not in the document]
Comment (alecchendev) [reply lightning/src/chain/channelmonitor.rs:3337]: Yea, I added this before I handled the duplicate HTLCUpdate events in ChannelManager - I guess maybe I should get rid of it since it's not really doing anything meaningful now?

File: lightning/src/ln/functional_tests.rs [hunk 5]
```diff
@@ -2211,6 +2211,118 @@ fn channel_reserve_in_flight_removes() {
...
+	let node_cfgs = create_node_cfgs(3, &chanmon_cfgs);
+	let node_chanmgrs = create_node_chanmgrs(3, &node_cfgs, &[None, None, None]);
+	let nodes = create_network(3, &node_cfgs, &node_chanmgrs);
+
+	create_announced_chan_between_nodes(&nodes, 0, 1);
+	let chan_2 = create_announced_chan_between_nodes(&nodes, 1, 2);
+
+	connect_blocks(&nodes[0], 2*CHAN_CONFIRM_DEPTH + 1 - nodes[0].best_block_info().1);
+	connect_blocks(&nodes[1], 2*CHAN_CONFIRM_DEPTH + 1 - nodes[1].best_block_info().1);
+	connect_blocks(&nodes[2], 2*CHAN_CONFIRM_DEPTH + 1 - nodes[2].best_block_info().1);
```
[the remark this answers is not in the document]
Comment (alecchendev) [reply lightning/src/ln/functional_tests.rs:2242]: Not totally necessary but having each node start on the same block made it easier to reason about the expiry times and how many blocks to connect to get exactly past the buffer, e.g. creating a channel automatically confirms some blocks on each involved node, so nodes[1] started ahead of the others, and then the route's cltv is formed relative to nodes[0]'s block height etc.

File: lightning/src/ln/functional_tests.rs [hunk 6]
```diff
@@ -2211,6 +2211,118 @@ fn channel_reserve_in_flight_removes() {
...
+			assert!(update_fail_malformed_htlcs.is_empty());
+			assert!(update_fee.is_none());
+			(update_fail_htlcs[0].clone(), commitment_signed.clone())
+		},
+		_ => panic!("Unexpected event"),
+	};
+
+	nodes[0].node.handle_update_fail_htlc(&nodes[1].node.get_our_node_id(), &update_fail);
+	commitment_signed_dance!(nodes[0], nodes[1], commitment_signed, false);
+	expect_payment_failed_conditions(&nodes[0], payment_hash, false, PaymentFailedConditions::new().blamed_chan_closed(true));
```
[the remark this answers is not in the document]
Comment (alecchendev) [reply lightning/src/ln/functional_tests.rs:2277]: Right, IIUC the `PaymentFailedConditions::new().blamed_chan_closed(true)` is more referring to that this payment failed because a downstream channel closed

File: lightning/src/ln/functional_tests.rs [hunk 7]
```diff
@@ -2211,6 +2211,118 @@ fn channel_reserve_in_flight_removes() {
...
+			connect_blocks(&nodes[2], TEST_FINAL_CLTV - CLTV_CLAIM_BUFFER + 2);
+			let node_2_txn = test_txn_broadcast(&nodes[2], &chan_2, None, HTLCType::SUCCESS);
+			check_closed_broadcast!(nodes[2], true);
+			check_closed_event(&nodes[2], 1, ClosureReason::CommitmentTxConfirmed, false);
+			check_added_monitors!(nodes[2], 1);
+
+			mine_transaction(&nodes[1], &node_2_txn[0]); // Commitment
+			mine_transaction(&nodes[1], &node_2_txn[1]); // HTLC success
+		},
+		PostFailBackAction::FailOffChain => {
```
[the remark this answers is not in the document]
Comment (alecchendev) [reply lightning/src/ln/functional_tests.rs:2301]: I think this is still tested since if after `nodes[1]` handled `nodes[2]`'s off-chain fail/claim, the test would fail if `nodes[1]` had an `update_fail_htlc` message waiting to be sent to `nodes[0]`? I could also explicitly assert this to make it more clear that this is being tested?

File: lightning/src/chain/channelmonitor.rs [hunk 8]
```diff
@@ -3303,6 +3312,60 @@ impl<Signer: WriteableEcdsaChannelSigner> ChannelMonitorImpl<Signer> {
...
+					// Only check forwarded HTLCs' previous hops
+					let source = match source_opt {
+						Some(source) => source,
+						None => continue,
+					};
+					let cltv_expiry = match source {
+						HTLCSource::PreviousHopData(HTLCPreviousHopData { cltv_expiry: Some(cltv_expiry), .. }) => *cltv_expiry,
+						_ => continue,
+					};
+					if cltv_expiry <= height + TIMEOUT_FAIL_BACK_BUFFER {
```
[the remark this answers is not in the document]
Comment (alecchendev) [reply lightning/src/chain/channelmonitor.rs:3332]: Makes sense. I'll try and check the other implementations to make sure before ditching the buffer, since the grace period seems to be only mentioned in an example in the spec as opposed to being explicitly required AFAICT

File: lightning/src/ln/channel.rs [hunk 9]
```diff
@@ -2193,6 +2193,18 @@ impl<Signer: WriteableEcdsaChannelSigner> Channel<Signer> {
 		self.context.channel_transaction_parameters.funding_outpoint.unwrap()
 	}
 
+	/// Returns whether an inbound HTLC on this channel has already been failed.
+	/// Returns None if the HTLC is not found.
+	pub fn check_inbound_htlc_failed(&self, htlc_id: u64) -> Option<bool> {
```
[the remark this answers is not in the document]
Comment (alecchendev) [reply lightning/src/ln/channel.rs:2198]: Although in general they're handled, in tests I think they should get caught by the debug assert in `Channel::fail_htlc/get_update_fulfill_htlc` asserting that if we are claiming/failing back an HTLC that's no longer present, it must be already claimed (i.e. present in `self.context.historical_inbound_htlc_fulfills`).

Seeing that I imagined we always prevented duplicate fails, though now that I'm looking there again I see the comment saying in case of block rescans we handle them. So, is it fine to remove that debug assert, or is there a way to test this that works around it somehow?

File: lightning/src/chain/channelmonitor.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (alecchendev) [reply lightning/src/chain/channelmonitor.rs:3337]: > Yeah I think we should.

Similar case to https://github.com/lightningdevkit/rust-lightning/pull/2457#discussion_r1294097707, am I just taking debug asserts too seriously..? There's another debug assert that just prevents us from failing the same HTLC twice in one process-pending-htlc cycle (which was what this code block covered), so should I just remove it since we actually handle them fine?

> We should also limit the amount of times these MonitorEvents will be queued somehow, as it seems we'll just continue to do so endlessly on every new block?

Oof yea. Added just a simple vec to keep track of HTLC ids we've failed already. Since we'll be force-closed by the time this is used, and (probably?) shouldn't have many HTLCs being failed back this way, I figured a vec was sufficient over a hash set or something else

**Comment by alecchendev** [comment]
Just pushed because I was sitting on some changes waiting till I could address anything - just going to add todos in the PR description. Also ignore https://github.com/lightningdevkit/rust-lightning/pull/2457/commits/2aad3247dcb2eec792df08dc20811f8efa75b8a9 pending me possibly just removing the debug asserts

File: lightning/src/chain/channelmonitor.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (alecchendev) [reply lightning/src/chain/channelmonitor.rs:3339]: > Maybe, but if they dont, there's also races across when we see blocks.

Are you saying this is in addition to the grace period? Isn't this what the grace period is for?

**Comment by alecchendev** [comment]
> While I definitely agree we should do this, I do wonder if we should gate it on HTLCs only being below some threshold, or only with some config knob. Its certainly surprising to users that we'd do this.

I was initially thinking we would always just want to keep the channel open because if we've reached our upstream timeout, we've lost that HTLC, and who knows how fast/slow their tx will confirm. But then again...we reached our own timeout downstream and we don't consider that our downstream counterparty has lost *that* HTLC just yet.

And if we do want to risk the channel for any HTLC, we probably only want to for larger ones, and not for smaller, so I'm leaning towards an amount threshold. Although determining a value for which a channel is worth closing depends on at least feerate (if we're the funder, how much we're paying on the commitment, and then how much it might be to open a new channel for this liquidity) so a fixed amount threshold might not be great?

File: lightning/src/chain/channelmonitor.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (alecchendev) [reply lightning/src/chain/channelmonitor.rs]: Got rid of the logic preventing generating a duplicate HTLCEvent (if we saw an onchain claim after doing our early fail back) especially now that we may not always fail back (based on the new config flag/threshold) based on this event, if we see the downstream claim, we should push that event to the channel manager.

> We should also limit the amount of times these MonitorEvents will be queued somehow, as it seems we'll just continue to do so endlessly on every new block?

> Added just a simple vec to keep track of HTLC ids we've failed already.

Though, now that whether we fail back an HTLC based on the new threshold could change with the feerate, should we still prevent regenerating these events on new blocks?

File: lightning/src/chain/channelmonitor.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (alecchendev) [reply lightning/src/chain/channelmonitor.rs]: Other implementations grace periods:
- CLN: [Looks like they have a grace period of 1 block](https://github.com/ElementsProject/lightning/blob/8d5be61fcb57f588908c7014de8fe31c4c5953cc/lightningd/peer_htlcs.c#L2710)
- LND: [Looks like they have a grace period denoted in time behind a config flag](https://github.com/lightningnetwork/lnd/blob/cbb4c27a444103eb27afecce47a97df2e6d738db/contractcourt/channel_arbitrator.go#L1673), though it [defaults to 0](https://github.com/lightningnetwork/lnd/blob/master/config.go#L68)
- Eclair: They don't have a grace period AFAICT? [Here](https://github.com/ACINQ/eclair/blob/3547f87f664c5e956a6d13af530a3d1cb6fc1052/eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala#L363C1-L363C1) they define expired htlcs, which is called when handling a new block [here](https://github.com/ACINQ/eclair/blob/master/eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala#L2576C3-L2576C3), which directly force closes [here](https://github.com/ACINQ/eclair/blob/3547f87f664c5e956a6d13af530a3d1cb6fc1052/eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ErrorHandlers.scala#L99)

So, I've left the buffer, but replaced it with our existing `LATENCY_GRACE_PERIOD_BLOCKS`

File: lightning/src/ln/channel.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (alecchendev) [reply lightning/src/ln/channel.rs]: I just ended up removing the debug asserts

**Comment by alecchendev** [comment]
Rebased because of conflicts, and added 3 new commits to the beginning to put this logic behind a config flag/threshold. Also simplified/added to tests a bit, and addressed the couple things about the grace period, debug asserts, and duplicate events (see other comments)

File: lightning/src/ln/channel.rs [hunk 10]
```diff
@@ -2100,6 +2100,45 @@ impl<SP: Deref> Channel<SP> where
...
+	{
+		let commitment_weight = if self.context.is_outbound() {
+			let keys = self.context.build_holder_transaction_keys(self.context.cur_holder_commitment_transaction_number);
+			let commitment_stats = self.context.build_commitment_transaction(self.context.cur_holder_commitment_transaction_number, &keys, true, true, logger);
+			commitment_stats.total_fee_sat
+		} else { 0 };
+
+		let htlc_weight = htlc_success_tx_weight(self.context.get_channel_type());
+		let p2wpkh_weight = 31 * 4;
+		let input_weight = 272; // Including witness, assuming it spends from a p2wpkh
```
Comment (alecchendev) [inline lightning/src/ln/channel.rs:2116]: By the way I got this number from [this post](https://bitcoin.stackexchange.com/questions/100159/what-is-the-size-and-weight-of-a-p2wpkh-input)

**Comment by alecchendev** [comment]
Small change: changed the collection to keep track of already early-failed htlcs in the channel monitor from a vec to a hashset--at first I thought a hashset might have some more memory overhead, but after some further reading it doesn't seem like it really would, and either way it shouldn't make a significant difference--I figured I should just use the one that makes more sense
